src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,467 @@
+    pub ack_idle_cleanup: bool,
+    pub max_idle_time: NonZeroU64,
+}
+
+impl Default for HecAcknowledgementsConfig {
+    fn default() -> Self {
+        Self {
+            max_pending_acks: NonZeroU64::new(10_000_000).unwrap(),
+            max_number_of_ack_channels: NonZeroU64::new(1_000_000).unwrap(),
+            max_pending_acks_per_channel: NonZeroU64::new(1_000_000).unwrap(),
+            ack_idle_cleanup: false,
+            max_idle_time: NonZeroU64::new(300).unwrap(),
+        }
+    }
+}
+
+pub struct IndexerAcknowledgement {
+    max_pending_acks: u64,
+    max_pending_acks_per_channel: u64,
+    max_number_of_ack_channels: u64,
+    channels: Arc<RwLock<HashMap<String, Arc<Channel>>>>,
+    shutdown: Shared<ShutdownSignal>,
+    total_pending_acks: AtomicU64,
+}
+
+impl IndexerAcknowledgement {
+    pub fn new(config: HecAcknowledgementsConfig, shutdown: Shared<ShutdownSignal>) -> Self {
+        let channels: Arc<RwLock<HashMap<String, Arc<Channel>>>> =
+            Arc::new(RwLock::new(HashMap::new()));
+        let max_idle_time = u64::from(config.max_idle_time);
+        let idle_task_channels = Arc::clone(&channels);
+
+        if config.ack_idle_cleanup {
+            tokio::spawn(async move {
+                let mut interval = interval(Duration::from_secs(max_idle_time));
+                loop {
+                    interval.tick().await;
+                    let mut channels = idle_task_channels.write().await;
+                    let expiring_channels = channels
+                        .iter()
+                        .filter_map(|(channel_id, channel)| {
+                            if channel.get_last_used().elapsed().as_secs() >= max_idle_time {
+                                Some(channel_id.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<String>>();
+
+                    for channel_id in expiring_channels {
+                        channels.remove(&channel_id);
+                    }
+                }
+            });
+        }
+
+        Self {
+            max_pending_acks: u64::from(config.max_pending_acks),
+            max_pending_acks_per_channel: u64::from(config.max_pending_acks_per_channel),
+            max_number_of_ack_channels: u64::from(config.max_number_of_ack_channels),
+            channels,
+            shutdown,
+            total_pending_acks: AtomicU64::new(0),
+        }
+    }
+
+    /// Creates a channel with the specified id if it does not exist.
+    async fn create_or_get_channel(&self, id: String) -> Result<Arc<Channel>, Rejection> {
+        let channels = self.channels.read().await;
+        if let Some(channel) = channels.get(&id) {
+            return Ok(Arc::clone(channel));
+        }
+        drop(channels);
+
+        let mut channels = self.channels.write().await;
+        if channels.len() < self.max_number_of_ack_channels as usize {
```

Comment:
This looks like it introduces a race condition. Two threads could call this with the same channel id and both reach this point and create the channel twice. I think you have to have a write lock at the outset in this function, which implies that it should just be a mutex.

Reply:
Gotcha. I changed this to a mutex.